java-checks/src/main/java/org/sonar/java/checks/spring/StatusCodesOnResponseCheck.java

```diff
@@ -0,0 +1,150 @@
+        checkTryCatchForStatus(methodInvocationTree);
+      }
+
+      MethodMatchers okMethodMatchersWithParam = MethodMatchers.create()
+        .ofTypes(RESPONSE_ENTITY)
+        .names("ok")
+        .withAnyParameters()
+        .build();
+
+      if (okMethodMatchersWithParam.matches(methodInvocationTree)) {
+        Tree catchParent = checkCatchWithErrorStatus(methodInvocationTree, false);
+
+        if (catchParent == null) {
+          checkTryWithOkStatus(methodInvocationTree, true);
+        }
+      }
+
+      MethodMatchers errorMethodMatchersWithParam = MethodMatchers.create()
+        .ofTypes(RESPONSE_ENTITY)
+        .names("badRequest", "notFound")
+        .addWithoutParametersMatcher()
+        .build();
+
+      if (errorMethodMatchersWithParam.matches(methodInvocationTree)) {
+        Tree catchParent = checkCatchWithErrorStatus(methodInvocationTree, true);
+
+        if (catchParent == null) {
+          checkTryWithOkStatus(methodInvocationTree, false);
+        }
+      }
+
+      super.visitMethodInvocation(methodInvocationTree);
+    }
+
+    private void checkTryCatchForStatus(MethodInvocationTree methodInvocationTree) {
+      Tree catchParent = ExpressionUtils.getParentOfType(methodInvocationTree, Tree.Kind.CATCH);
+      boolean isError = methodInvocationTree.arguments().stream()
+        .map(MemberSelectExpressionTree.class::cast)
+        .anyMatch(arg -> "INTERNAL_SERVER_ERROR".equals(arg.identifier().name()) || "NOT_FOUND".equals(arg.identifier().name()));
+
+      if (catchParent != null && !isError) {
+        reportIssue(methodInvocationTree,
+          "Use the \"ResponseEntity.badRequest()\" or \"ResponseEntity.notFound()\" method" +
+            "or set the status to \"HttpStatus.INTERNAL_SERVER_ERROR\" or \"HttpStatus.NOT_FOUND\".");
+      }
+
+      if (catchParent == null) {
+        Tree tryParent = ExpressionUtils.getParentOfType(methodInvocationTree, Tree.Kind.TRY_STATEMENT);
+        boolean isOk = methodInvocationTree.arguments().stream()
+          .map(MemberSelectExpressionTree.class::cast)
+          .anyMatch(arg -> "OK".equals(arg.identifier().name()));
+
+        if (tryParent != null && !isOk) {
+          reportIssue(methodInvocationTree, "Use the \"ResponseEntity.ok()\" method or set the status to \"HttpStatus.OK\".");
+        }
+      }
+    }
+
+    private Tree checkCatchWithErrorStatus(MethodInvocationTree methodInvocationTree, boolean isError) {
+      Tree catchParent = ExpressionUtils.getParentOfType(methodInvocationTree, Tree.Kind.CATCH);
+
+      if (catchParent != null && !isError) {
+        reportIssue(methodInvocationTree,
+          "Use the \"ResponseEntity.badRequest()\" or \"ResponseEntity.notFound()\" method" +
+            "or set the status to \"HttpStatus.INTERNAL_SERVER_ERROR\" or \"HttpStatus.NOT_FOUND\".");
+      }
+      return catchParent;
+    }
+
+    private void checkTryWithOkStatus(MethodInvocationTree methodInvocationTree, boolean isOk) {
```

Comment:
I am a bit confused as to why two methods with similar signatures `check[Catch|Try]with[Error|Ok]Status` sometimes return a Tree sometimes. Can we align on that?